### PR TITLE
JMS: Support message individual timeToLive during send (#441)

### LINKED ISSUE #441: JMS: Support individual  time-to-live of jms messages in JmsSinkStage during sending
At the moment exists only the possibility to set a default timeToLive for messages at the jms producer. I would like to enhance this behavior to support an individual time-to-live. This can be done with the method 

` jmsProducer.send(message, deliveryMode,priority, timeToLiveInMillis)`

In order to pass the time-to-live to the JmsSinkStage i would like to enhance the JmsTextMessage with this information. If I am doing this, there would be  merge conflicts with the pull request [#199](https://github.com/akka/alpakka/pull/430), where the JmsMessages are refactored out in a separat trait.

I can avoid this by doing this change in that pull request or separat in another. Which one you prefer?

### PR DESCRIPTION
**Pull request by d2e8f9**
Add support for attributes timeToLive, priority and deliveryMode per message in JmsSinkStage.

Resolves #441

### CODE REVIEW

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala [hunk 1]
```diff
@@ -7,10 +7,34 @@ package akka.stream.alpakka.jms
...
-sealed trait JmsHeader
-final case class JmsCorrelationId(jmsCorrelationId: String) extends JmsHeader
-final case class JmsReplyTo(jmsDestination: Destination) extends JmsHeader
-final case class JmsType(jmsType: String) extends JmsHeader
+sealed trait JmsHeader {
+
+  /**
+   * Indicates if this header must be set during the send() operation according to the JMS specification or as attribute of the jms message before.
+   */
+  def usedDuringSend(): Boolean
```
Comment (ennru) [inline jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: you may remove the ()

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala [hunk 2]
```diff
@@ -7,10 +7,34 @@ package akka.stream.alpakka.jms
...
+}
+
+final case class JmsCorrelationId(jmsCorrelationId: String) extends JmsHeader {
+  override def usedDuringSend(): Boolean = false
```
Comment (ennru) [inline jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: when declared without the () in the trait, you may override it with val

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala [hunk 3]
```diff
@@ -41,6 +65,30 @@ object JmsType {
   def create(jmsType: String) = JmsType(jmsType)
 }
 
+object JmsTimeToLive {
+
+  /**
+   * Java API: create  [[JmsTimeToLive]]
+   */
+  def create(timeInMillis: Long) = JmsTimeToLive(timeInMillis)
```
Comment (ennru) [inline jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: you may add an `apply` method taking `scala.concurrent.duration.Duration` and possibly a Java version with `TimeUnit`

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala [hunk 4]
```diff
@@ -7,10 +7,34 @@ package akka.stream.alpakka.jms
...
+  override def usedDuringSend(): Boolean = false
+}
+
+final case class JmsTimeToLive(timeInMillis: Long) extends JmsHeader {
+  override def usedDuringSend(): Boolean = true
+}
+final case class JmsPriority(priority: Int) extends JmsHeader {
+  override def usedDuringSend(): Boolean = true
+}
+final case class JmsDeliveryMode(deliveryMode: Int) extends JmsHeader {
```
Comment (ennru) [inline jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: Add some Scaladoc with reference to `javax.jms.DeliveryMode`

**Comment by d2e8f9** [comment]
Sorry. I made a mistake during the rebase with master and applying the code review changes. So the git history is rewritten. I will fix that.

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala (hunk 1 above)
Comment (d2e8f9) [reply jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: Fixed

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala (hunk 2 above)
Comment (d2e8f9) [reply jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: Fixed

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala (hunk 4 above)
Comment (d2e8f9) [reply jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: Fixed. Please take a look

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala (hunk 3 above)
Comment (d2e8f9) [reply jms/src/main/scala/akka/stream/alpakka/jms/JmsMessages.scala]: Fixed

File: jms/src/test/java/akka/stream/alpakka/jms/javadsl/JmsConnectorsTest.java [hunk 5]
```diff
@@ -332,6 +334,9 @@ public void publishAndConsumeJmsTextMessagesWithHeaders() throws Exception {
                     .map(jmsTextMessage -> jmsTextMessage.withHeader(JmsType.create("type")))
                     .map(jmsTextMessage -> jmsTextMessage.withHeader(JmsCorrelationId.create("correlationId")))
                     .map(jmsTextMessage -> jmsTextMessage.withHeader(JmsReplyTo.queue("test-reply")))
+                    .map(jmsTextMessage -> jmsTextMessage.withHeader(JmsTimeToLive.create(999, TimeUnit.SECONDS)))
```
Comment (ennru) [inline jms/src/test/java/akka/stream/alpakka/jms/javadsl/JmsConnectorsTest.java]: Not really changed in this PR, but wouldn't it be much better to chain all `withHeader` calls in one map operation?

Comment (d2e8f9) [reply jms/src/test/java/akka/stream/alpakka/jms/javadsl/JmsConnectorsTest.java]: Of course. Fixed it and added documentation of sending messages with headers.

**Comment by ennru** [comment]
Look good to me.

**Comment by ennru** [comment]
Thank you for this contribution. Keep them coming!

**Pull request merged**